README.md
# Franklin Chat UI
The Franklin Chat UI is a React app that connects to the Franklin Chat Service and allows users to chat with the Franklin team.

## Using the app
- Open the app in a browser by going to https://main--franklin-chat-ui--adobe.hlx.live/index.html
- Enter your business email address and click on the "Login" button
- Open the email you received and click on the link
- Go back to the app and start chatting

## Development

To build and deploy the app, you need to create a `.env` file in the root directory with the following properties:
```
REACT_APP_AWS_SERVICE_ENDPOINT=ws://localhost:8081
REACT_APP_MAGIC_LINK_API_KEY=...
PORT=8080
```

In the project directory, you can run:

### `npm start`

Runs the app in the development mode.\
Open [http://localhost:3001](http://localhost:3001) to view it in the browser.

The page will reload if you make edits.\
You will also see any lint errors in the console.

### `npm test`

Launches the test runner in the interactive watch mode.\
See the section about [running tests](https://facebook.github.io/create-react-app/docs/running-tests) for more information.

### `npm run build`

Builds the app for production to the `build` folder.\
It correctly bundles React in production mode and optimizes the build for the best performance.

The build is minified and the filenames include the hashes.\
Your app is ready to be deployed!

See the section about [deployment](https://facebook.github.io/create-react-app/docs/deployment) for more information.

### Cache flushing

The project contains a folder mapping from `/` to `/build/index.html`. When releasing a new version, `/build/index.html` gets updated, https://main--franklin-chat-ui--adobe.hlx.live/build/index.html is flushed. There seems to be a limitation and the folder mapping does not get updated automatically, i.e. https://main--franklin-chat-ui--adobe.hlx.live/ will still serve the old version. To flush the folder mapping, you can run:

```
curl -X POST https://admin.hlx.page/code/adobe/franklin-chat-ui/main/*
```

CHANGELOG.md
## [2.3.1](https://github.com/adobe/franklin-chat-ui/compare/v2.3.0...v2.3.1) (2023-11-15)


### Bug Fixes

* adjust success cmd ([f933ee9](https://github.com/adobe/franklin-chat-ui/commit/f933ee907b704b7c62d353e441e55ac02f0db68c))

# [2.3.0](https://github.com/adobe/franklin-chat-ui/compare/v2.2.0...v2.3.0) (2023-11-15)


### Features

* version check ([#39](https://github.com/adobe/franklin-chat-ui/issues/39)) ([c06109e](https://github.com/adobe/franklin-chat-ui/commit/c06109efeec9acf018be6f48da427395bd00e39b))

# [2.2.0](https://github.com/adobe/franklin-chat-ui/compare/v2.1.0...v2.2.0) (2023-11-13)


### Features

* close client when no mapping found ([#37](https://github.com/adobe/franklin-chat-ui/issues/37)) ([2573d4f](https://github.com/adobe/franklin-chat-ui/commit/2573d4f19fdcc7d678349aecc484a0b5f7db8330))

# [2.1.0](https://github.com/adobe/franklin-chat-ui/compare/v2.0.1...v2.1.0) (2023-10-25)


### Features

* send client version ([#36](https://github.com/adobe/franklin-chat-ui/issues/36)) ([9f8534d](https://github.com/adobe/franklin-chat-ui/commit/9f8534de8b4e9de05de557ee551b6936513f2b01))

## [2.0.1](https://github.com/adobe/franklin-chat-ui/compare/v2.0.0...v2.0.1) (2023-10-24)


### Bug Fixes

* **index:** update app title ([#35](https://github.com/adobe/franklin-chat-ui/issues/35)) ([439f7a8](https://github.com/adobe/franklin-chat-ui/commit/439f7a84bab73f1d50e0bfc48d804deac7a2ba25))

# [2.0.0](https://github.com/adobe/franklin-chat-ui/compare/v1.22.0...v2.0.0) (2023-10-20)


### Bug Fixes

* major release ([af77457](https://github.com/adobe/franklin-chat-ui/commit/af77457fde39637696e7b3f14a26b68da6661ebc))
* remove branch from release ([#34](https://github.com/adobe/franklin-chat-ui/issues/34)) ([ada2b3e](https://github.com/adobe/franklin-chat-ui/commit/ada2b3e9865c3b8c15e09f155c4d5c5216341d99))
* trigger release ([1017a22](https://github.com/adobe/franklin-chat-ui/commit/1017a22e41fef0fe05d5a46d0bca53e39531778a))
* trigger release\nBREAKING CHANGE: major re-architecture\n ([8eff9b1](https://github.com/adobe/franklin-chat-ui/commit/8eff9b15255c7c40024e30ddef23f71c0ab939bc))
* version mess ([#33](https://github.com/adobe/franklin-chat-ui/issues/33)) ([fc891d1](https://github.com/adobe/franklin-chat-ui/commit/fc891d177424d9f91a8bc37863a3809641c67659))


### Features

* switch to aws backend ([#19](https://github.com/adobe/franklin-chat-ui/issues/19)) ([fe09eca](https://github.com/adobe/franklin-chat-ui/commit/fe09eca4bcf13dd7320c5d99ec5db159fa3026b3))
* trigger major release attempt ([f83d9b8](https://github.com/adobe/franklin-chat-ui/commit/f83d9b805d800bee24dd9b6b0cbab4c36ca35183))


### BREAKING CHANGES

* switch to aws backend
* switch to aws backend

# [1.25.0](https://github.com/adobe/franklin-chat-ui/compare/v1.24.0...v1.25.0) (2023-10-20)


### Features

* display current version ([1504c10](https://github.com/adobe/franklin-chat-ui/commit/1504c105756d36abf64ce515ec7c2482f4d592a8)), closes [#30](https://github.com/adobe/franklin-chat-ui/issues/30)

# [1.24.0](https://github.com/adobe/franklin-chat-ui/compare/v1.23.0...v1.24.0) (2023-10-05)


### Features

* aws endpoint secret ([3fb8f6c](https://github.com/adobe/franklin-chat-ui/commit/3fb8f6cf242f48b0d007fd87b56ad108d5760fd2))

# [1.23.0](https://github.com/adobe/franklin-chat-ui/compare/v1.22.0...v1.23.0) (2023-10-03)


### Bug Fixes

* timeout cleanup ([da74169](https://github.com/adobe/franklin-chat-ui/commit/da74169a759e839f231735f0eac650042ddfb9da))


### Features

* add aws branch to semantic release ([020136d](https://github.com/adobe/franklin-chat-ui/commit/020136dd00725d6a4f8277b12a1b81e4a704294f))
* add aws branch to semantic release ([68f6026](https://github.com/adobe/franklin-chat-ui/commit/68f60261efe4737b197cc81d922f169123448bbf))
* better accessibility handling ([dcea29d](https://github.com/adobe/franklin-chat-ui/commit/dcea29da90cdbce8a9610f550d7c1f9595906cdf))
* connect to aws service ([e6c15de](https://github.com/adobe/franklin-chat-ui/commit/e6c15de6a7e7ff5420478df16c49569ed40a2550))

# [1.22.0](https://github.com/adobe/franklin-chat-ui/compare/v1.21.1...v1.22.0) (2023-09-21)


### Features

* improve weight ([7a070a8](https://github.com/adobe/franklin-chat-ui/commit/7a070a883606819d52c35ed75f17f5e9e8889d8c))

## [1.21.1](https://github.com/adobe/franklin-chat-ui/compare/v1.21.0...v1.21.1) (2023-06-30)


### Bug Fixes

* trigger release ([7c102ef](https://github.com/adobe/franklin-chat-ui/commit/7c102ef03a2d7860329dcca434962089ddcbbd64))

# [1.21.0](https://github.com/adobe/franklin-chat-ui/compare/v1.20.1...v1.21.0) (2023-06-30)


### Bug Fixes

* attempt to fix release versioning ([fff016d](https://github.com/adobe/franklin-chat-ui/commit/fff016d5cf8d84f812587edc3b16e77c4b95b511))
* trigger release ([1bb8e9e](https://github.com/adobe/franklin-chat-ui/commit/1bb8e9eb0d1dd88396c9f735a232a106584db4fe))
* trigger release ([e18a7c3](https://github.com/adobe/franklin-chat-ui/commit/e18a7c3fc8e695b08cc181e9e692921b94238106))
* trigger release ([6731fe8](https://github.com/adobe/franklin-chat-ui/commit/6731fe8e4c8293e93d4f4e13652e3e75609b7eb8))
* trigger release ([53fa0ad](https://github.com/adobe/franklin-chat-ui/commit/53fa0ad0fe00a2f733e9e28c2306ef25dd0d2e3d))


### Features

* build as homepage ([#23](https://github.com/adobe/franklin-chat-ui/issues/23)) ([947d0e5](https://github.com/adobe/franklin-chat-ui/commit/947d0e5d3e6acdd25f0145e3f1bd9e8b9efaffdb))
* collect rum data ([#25](https://github.com/adobe/franklin-chat-ui/issues/25)) ([aee4e8b](https://github.com/adobe/franklin-chat-ui/commit/aee4e8b36a61aa8b2d6eb5e4a820ee013b24d66c))

# [1.21.0](https://github.com/adobe/franklin-chat-ui/compare/v1.20.1...v1.21.0) (2023-06-30)


### Bug Fixes

* attempt to fix release versioning ([fff016d](https://github.com/adobe/franklin-chat-ui/commit/fff016d5cf8d84f812587edc3b16e77c4b95b511))

src/rum.js
/**
 * log RUM if part of the sample.
 * @param {string} checkpoint identifies the checkpoint in funnel
 * @param {Object} data additional data for RUM sample
 */
export function sampleRUM(checkpoint, data = {}) {
  sampleRUM.defer = sampleRUM.defer || [];
  const defer = (fnname) => {
    sampleRUM[fnname] = sampleRUM[fnname]
      || ((...args) => sampleRUM.defer.push({ fnname, args }));
  };
  sampleRUM.drain = sampleRUM.drain
    || ((dfnname, fn) => {
      sampleRUM[dfnname] = fn;
      sampleRUM.defer
        .filter(({ fnname }) => dfnname === fnname)
        .forEach(({ fnname, args }) => sampleRUM[fnname](...args));
    });
  sampleRUM.on = (chkpnt, fn) => { sampleRUM.cases[chkpnt] = fn; };
  defer('observe');
  defer('cwv');
  try {
    window.hlx = window.hlx || {};
    if (!window.hlx.rum) {
      // const usp = new URLSearchParams(window.location.search);
      // const weight = (usp.get('rum') === 'on') ? 1 : 10; // with parameter, weight is 1. Defaults to 10.
      const weight = 10; // temporary until we have enough users
      // eslint-disable-next-line no-bitwise
      const hashCode = (s) => s.split('').reduce((a, b) => (((a << 5) - a) + b.charCodeAt(0)) | 0, 0);
      const id = `${hashCode(window.location.href)}-${new Date().getTime()}-${Math.random().toString(16).substr(2, 14)}`;
      const random = Math.random();
      const isSelected = (random * weight < 1);
      const urlSanitizers = {
        full: () => window.location.href,
        origin: () => window.location.origin,
        path: () => window.location.href.replace(/\?.*$/, ''),
      };
      // eslint-disable-next-line object-curly-newline, max-len
      window.hlx.rum = { weight, id, random, isSelected, sampleRUM, sanitizeURL: urlSanitizers[window.hlx.RUM_MASK_URL || 'path'] };
    }
    const { weight, id } = window.hlx.rum;
    if (window.hlx && window.hlx.rum && window.hlx.rum.isSelected) {
      const sendPing = (pdata = data) => {
        // eslint-disable-next-line object-curly-newline, max-len, no-use-before-define
        const body = JSON.stringify({ weight, id, referer: window.hlx.rum.sanitizeURL(), checkpoint, ...data });
        const url = `https://rum.hlx.page/.rum/${weight}`;
        // eslint-disable-next-line no-unused-expressions
        navigator.sendBeacon(url, body);
        // eslint-disable-next-line no-console
        console.debug(`ping:${checkpoint}`, pdata);
      };
      sampleRUM.cases = sampleRUM.cases || {
        cwv: () => sampleRUM.cwv(data) || true,
        lazy: () => {
          // use classic script to avoid CORS issues
          const script = document.createElement('script');
          script.src = 'https://rum.hlx.page/.rum/@adobe/helix-rum-enhancer@^1/src/index.js';
          document.head.appendChild(script);
          return true;
        },
      };
      sendPing(data);
      if (sampleRUM.cases[checkpoint]) { sampleRUM.cases[checkpoint](); }
    }
  } catch (error) {
    // something went wrong
  }
}

src/Utils.ts
const EMOJIS: { [key: string]: string } = {
  ':smile:': '😄',
  ':laughing:': '😆',
  ':wink:': '😉',
  ':heart:': '❤️',
  ':thumbsup:': '👍',
  ':thumbsdown:': '👎',
  ':confused:': '😕',
  ':sob:': '😭',
  ':sunglasses:': '😎',
  ':thinking_face:': '🤔',
  ':sweat_smile:': '😅',
  ':cry:': '😢',
  ':astonished:': '😲',
  ':joy:': '😂',
  ':rolling_on_the_floor_laughing:': '🤣',
  ':smirk:': '😏',
  ':neutral_face:': '😐',
  ':expressionless:': '😑',
  ':no_mouth:': '😶',
  ':grin:': '😁',
  ':clap:': '👏',
  ':wave:': '👋',
  ':tada:': '🎉',
  ':fire:': '🔥',
  ':100:': '💯',
  ':raised_hands:': '🙌',
  ':pray:': '🙏',
  ':muscle:': '💪',
  ':sparkles:': '✨',
  ':ok_hand:': '👌',
  ':+1:': '👍',
  ':-1:': '👎',
  ':facepalm:': '🤦',
  ':shrug:': '🤷',
  ':eyes:': '👀',
  ':tongue:': '😛',
  ':see_no_evil:': '🙈',
  ':hear_no_evil:': '🙉',
  ':speak_no_evil:': '🙊',
  ':mask:': '😷',
  ':alien:': '👽',
  ':ghost:': '👻',
  ':star:': '⭐',
  ':zap:': '⚡',
  ':umbrella:': '☔',
  ':snowflake:': '❄️',
  ':checkered_flag:': '🏁',
  ':coffee:': '☕',
  ':birthday:': '🎂',
  ':scream:': '😱',
  ':rage:': '😡',
  ':v:': '✌️',
  ':warning:': '⚠️',
  ':white_check_mark:': '✅',
  ':heavy_exclamation_mark:': '❗',
  ':question:': '❓',
  ':slightly_smiling_face:': '🙂',
};

const APP_VERSION = process.env.REACT_APP_VERSION || 'unknown';
console.log(`App version: ${APP_VERSION}`);

export function getAppVersion() {
  return APP_VERSION;
}

export async function checkVersion() {
  try {
    const url = new URL(window.location.href);
    if (url.searchParams.has('ck')) {
      url.searchParams.delete('ck');
      window.history.replaceState({}, '', url.toString());
    } else if (!url.hostname.startsWith('localhost')) {
      // do not reload if running locally and if ck is already present to avoid infinite reload loop
      // next join will try again
      const res = await fetch('/package.json');
      if (res.ok) {
        const json = await res.json();
        if (json.version !== APP_VERSION) {
          console.log('You are running an old version of the app. We are reloading your browser.');
          url.searchParams.set('ck', new Date().getTime().toString());
          window.location.replace(url.toString()); 
        }
      }
    }
  } catch (e) {
    console.error('Error in version checker',e);
  }
}

export const REVERSED_EMOJIS: { [key: string]: string } = Object.entries(EMOJIS).reduce(
  (acc, [key, value]) => {
    acc[value] = key;
    return acc;
  },
  {} as { [key: string]: string }
);

export function convertSlackToHtml(slackText: string, teamId: string) {
  slackText = convertSlackLinksToHTML(slackText);
  slackText = convertHereMention(slackText);
  slackText = convertSlackUserMentionsToDeepLinks(slackText, teamId);
  slackText = convertAllEmojiToUnicode(slackText);
  slackText = slackText.replace(/\*(.*?)\*/g, '<strong>$1</strong>');
  slackText = slackText.replace(/_(.*?)_/g, '<em>$1</em>');
  slackText = slackText.replace(/~(.*?)~/g, '<del>$1</del>');
  slackText = slackText.replace(/\n/g, '<br>');
  return slackText;
}

export function convertSlackTimestampToUTC(slackTimestamp: string) {
  const seconds = parseInt(slackTimestamp.split('.')[0]);
  const milliseconds = parseInt(slackTimestamp.split('.')[1]);
  const date = new Date(seconds * 1000 + milliseconds);
  return date.toUTCString();
}

function convertHereMention(message: string): string {
  return message.replaceAll('<!here>', '<b>@here</b>');
}

function convertSlackUserMentionsToDeepLinks(message: string, teamId: string): string {
  const slackUserMentionRegex = /<@(\w+)\|([^>]+)>/g;
  return message.replace(slackUserMentionRegex, (match, userId, userName) => {
    return `<b>@${userName}</b>`;
  });
}

function convertSlackLinksToHTML(text: string): string {
  const regex = /<(?![@#!])([^|>]+)(?:\|([^>]+))?>/g;
  return text.replace(regex, (_, url, linkText) => {
    if (!linkText) {
      linkText = url;
    }
    return `<a href="${url}">${linkText}</a>`;
  });
}

export function convertAllEmojiToUnicode(slackText: string) {
  for (const emoji in EMOJIS) {
    slackText = slackText.replace(emoji, EMOJIS[emoji]);
  }
  return slackText;
}

export function convertAllUnicodeToEmoji(text: string): string {
  let newText = text;
  for (const [emoji, code] of Object.entries(REVERSED_EMOJIS)) {
    newText = newText.split(emoji).join(code);
  }
  return newText;
}


src/ChatClient.ts
import {ConstantBackoff, Websocket, WebsocketBuilder,WebsocketEvents} from 'websocket-ts';
import { v4 as uuid } from 'uuid';
import {sampleRUM} from "./rum";
import {getAppVersion, checkVersion} from "./Utils";

export type User = {
  name: string,
  icon?: string,
};

export type Message = {
  ts: string
  user: User
  text: string
  threadId?: string
  replyCount?: number
  files?: Attachment[]
}

export type Attachment = {
  id: string
  name: string
  url: string
  thumbUrl: string
}

export enum ConnectionStatus {
  CONNECTED,
  DISCONNECTED,
}

export type ConnectionCallback = (status: ConnectionStatus) => void;
export type MessageCallback = (history: Message[]) => void;
export type ErrorCallback = (error: Error) => void;

const SERVICE_ENDPOINT = process.env.REACT_APP_AWS_SERVICE_ENDPOINT || 'ws://localhost:8081';
console.log(`Using service endpoint: ${SERVICE_ENDPOINT}`);

export class ChatClient {
  private client: Websocket;

  private email: string | undefined;
  private teamId: string | undefined;
  private channelId: string | undefined;
  private channelName: string | undefined;

  private history: Message[] = [];

  private connectionStatus: ConnectionStatus = ConnectionStatus.DISCONNECTED;
  private error: Error | undefined;

  private statusCallbacks: ConnectionCallback[] = [];
  private messageCallbacks: MessageCallback[] = [];
  private errorCallbacks: ErrorCallback[] = [];

  constructor(private token: string) {
    const builder = new WebsocketBuilder(`${SERVICE_ENDPOINT}?token=${token}`)
      .withBackoff(new ConstantBackoff(1000))
      .onClose((i, ev) => {
        this.fireStatusChange(ConnectionStatus.DISCONNECTED)
      })
      .onError((i, ev) => {
        console.log('Websocket error', ev);
        sampleRUM('chat:error', { source: 'client#socket', target: ev.target });
        this.fireError(new Error('Websocket error'));
      })
      .onRetry((i, ev) => {
        console.log("retry")
      })

    builder.onOpen(async (i, ev) => {
      await this.join();
    });

    builder.onMessage(async (i, ev) => {
      const message = JSON.parse(ev.data);
      switch (message.type) {
        case 'message':
          console.log(`Client message: ${JSON.stringify(message.data)}`);
          this.fireMessage(message.data)
          break;
        default:
          console.log(`Client unknown message: ${JSON.stringify(message)}`);
      }
    });

    this.client = builder.build();
  }

  addStatusCallback(callback: ConnectionCallback) {
    callback(this.connectionStatus);
    this.statusCallbacks.push(callback);
  }

  removeStatusCallback(callback: ConnectionCallback) {
    const index = this.statusCallbacks.indexOf(callback);
    if (index > -1) {
      this.statusCallbacks.splice(index, 1);
    }
  }

  addMessageCallback(callback: MessageCallback) {
    if (this.history.length > 0) {
      callback(this.history);
    }
    this.messageCallbacks.push(callback);
  }

  removeMessageCallback(callback: MessageCallback) {
    const index = this.messageCallbacks.indexOf(callback);
    if (index > -1) {
      this.messageCallbacks.splice(index, 1);
    }
  }

  addErrorCallback(callback: ErrorCallback) {
    if (this.error) {
      callback(this.error);
    }
    this.errorCallbacks.push(callback);
  }

  removeErrorCallback(callback: ErrorCallback) {
    const index = this.errorCallbacks.indexOf(callback);
    if (index > -1) {
      this.errorCallbacks.splice(index, 1);
    }
  }

  getConnectionStatus() {
    return this.connectionStatus;
  }

  getTeamId() {
    return this.teamId;
  }

  getChannelId() {
    return this.channelId;
  }

  getChannelName() {
    return this.channelName;
  }

  async join() {
    try {
      // check client version before joining
      await checkVersion();
      const {email, channelId, teamId, channelName} = await this.sendCommand<any, any>('join', {
        version: getAppVersion(),
      });
      if (!email || !channelId || !teamId || !channelName) {
        console.error(`Client ready without email or channel. Disconnecting...`);
        this.client.close();
        this.fireError(new Error('No mapping found.'));
        return;
      }

      this.email = email;
      this.teamId = teamId;
      this.channelId = channelId;
      this.channelName = channelName;

      console.log(`Client ready: ${email}, ${channelId}`);
      this.fireStatusChange(ConnectionStatus.CONNECTED);

      await this.requestHistory();
    } catch (e: any) {
      console.error(`Impossible to join. Disconnecting...`);
      this.client.close();
      this.fireError(new Error(e.message));
      return;
    }
  }

  async postMessage(text: string, ts?: string) {
    return this.sendCommand<any, any>('post', {
      threadId: ts,
      user: {
        name: this.email,
      },
      text
    });
  }

  async requestHistory() {
    const latest = this.history.length > 0 ? this.history[this.history.length - 1].ts : undefined;
    console.log(`requesting history, latest: ${latest}`);
    const messages: Message[] = await this.sendCommand<Message[], any>('history', { latest });
    messages.forEach((message: any) => {
      this.history.push(message);
    });
    console.log(`received ${messages.length} messages`);
    this.messageCallbacks.forEach(callback => callback(this.history));
    return messages.length > 0;
  }

  async getReplies(ts: string) {
    return this.sendCommand<Message[], any>('replies', { ts });
  }

  private sendCommand<T, P>(command: string, data: P, timeout: number = 5000): Promise<T> {
    return new Promise<T>((resolve, reject) => {
      const correlationId = uuid();
      console.log(`sending command ${command} with correlationId ${correlationId}`);
      let handle = null as any;
      const callback = (i: Websocket, ev: MessageEvent) => {
        console.log(`received message with correlationId ${correlationId}`);
        if (handle) {
          clearTimeout(handle);
        }
        const data: any = JSON.parse(ev.data);
        if (data.correlationId === correlationId) {

          console.log(`received response for command ${command}`);
          i.removeEventListener(WebsocketEvents.message, callback);

          if (data.error) {
            reject(new Error(data.error));
          } else {
            resolve(data.data);
          }
        }
      };

      handle = setTimeout(() => {
        console.warn(`timeout for command ${command}`);
        this.client.removeEventListener(WebsocketEvents.message, callback);
        reject(new Error('timeout'));
      }, timeout);

      this.client.addEventListener(WebsocketEvents.message, callback);

      this.client.send(JSON.stringify({
        type: command,
        correlationId,
        data
      }));

      sampleRUM('chat:client', { source: 'message', target: this.channelId });
    });
  }

  private fireError(error: Error) {
    this.error = error;
    this.errorCallbacks.forEach(callback => callback(error));
  }

  private fireStatusChange(status: ConnectionStatus) {
    this.connectionStatus = status;
    this.statusCallbacks.forEach(callback => callback(status));
  }

  private fireMessage(message: Message) {
    console.log("message: ", message);
    if (message.threadId) {
      const parent = this.history.find(m => m.ts === message.threadId);
      if (parent) {
        parent.replyCount = parent.replyCount ? parent.replyCount + 1 : 1;
      }
    } else {
      this.history.unshift(message);
    }
    this.messageCallbacks.forEach(callback => callback(this.history));
  }
}
